### PR TITLE
refactor(SG): make the settings system lighter and easier

### PR DESCRIPTION
**Pull request by kyranet**
### Description of the PR

Removed, removed, removed, and more removed. PR requested by @bdistin and approved by @UnseenFaith and me.

image

### Changes Proposed in this Pull Request (List new items in CHANGELOG.MD)

- Removed `Schema#path`.
- Removed `Schema#type`.
- Removed callback support in `Schema#add`.
- Removed custom `Schema#keys`.
- Removed custom `Schema#values`.
- Removed custom `Schema#entries`.
- Removed custom `Schema#@@iterator`.
- Removed `Schema.is`.
- Removed `SchemaEntry#path`, use `SchemaEntry#key` now.
- Removed `SchemaEntry.is`.
- Removed `SchemaFolder`.
- Removed custom `Settings#keys`.
- Removed custom `Settings#values`.
- Removed custom `Settings#entries`.
- Removed custom `Settings#@@iterator`.
- Removed `onlyConfigurable` option in `Settings`'s update methods.
- Removed `SettingsFolder`.
- Removed `GatewayStorage`, its code has been merged into `Gateway`.
- Removed so many things we even removed a lot of tests.

### Semver Classification

- [ ] This PR only includes documentation or non-code changes.
- [ ] This PR fixes a bug and does not change the (intended) framework interface.
- [ ] This PR adds methods or properties to the framework interface.
- [x] This PR removes or renames methods or properties in the framework interface.

### CODE REVIEW

File: src/lib/settings/Settings.ts [hunk 1]
```diff
@@ -76,11 +79,653 @@ export class Settings extends SettingsFolder {
...
+	 * await message.guild.settings.reset({ prefix: null });
+	 *
+	 * @example
+	 * // Resets multiple entries with a regular object:
+	 * await message.guild.settings.reset({ prefix: null, administrator: null });
+	 */
+	public async reset(object: ReadonlyKeyedObject, options?: Readonly<SettingsResetOptions>): Promise<SettingsUpdateResults>;
+	public async reset(paths: string | ReadonlyKeyedObject | readonly string[] = [...this.keys()], options: Readonly<SettingsResetOptions> = {}): Promise<SettingsUpdateResults> {
+		if (this.existenceStatus === SettingsExistenceStatus.Unsynchronized) {
+			throw new Error('Cannot reset keys from a pending to synchronize settings instance. Perhaps you want to call `sync()` first.');
```
Comment (vladfrangu) [inline src/lib/settings/Settings.ts:170]: Maybe:

```suggestion
			throw new Error('Cannot reset keys from an unsynchronized settings instance. Perhaps you want to call `sync()` first.');
```

File: src/lib/settings/Settings.ts [hunk 2]
```diff
@@ -76,11 +79,653 @@ export class Settings extends SettingsFolder {
...
+		}
+
+		return this._processUpdate(entries, options as InternalRawUpdateOptions);
+	}
+
+	/**
+	 * Overload to serialize this entry to JSON.
+	 */
+	public toJSON(): SettingsJson {
+		return Object.fromEntries([...super.entries()].map(([key, value]) => [key, value]));
```
Comment (vladfrangu) [inline src/lib/settings/Settings.ts:292]: Why is that map...Isn't that completely redundant?

**Comment by bdistin** [comment]
I apologize for that last commit, that should have been on master 😓

File: src/lib/settings/gateway/Gateway.ts [hunk 3]
```diff
@@ -1,10 +1,28 @@
 import { RequestHandler, IdKeyed } from '@klasa/request-handler';
-import { GatewayStorage, GatewayStorageOptions } from './GatewayStorage';
 import { Settings } from '../Settings';
 import { Cache } from '@klasa/cache';
+import { Schema } from '../schema/Schema';
```
Comment (vladfrangu) [inline src/lib/settings/gateway/Gateway.ts:4]: Import order

File: src/lib/settings/gateway/Gateway.ts [hunk 4]
```diff
@@ -18,8 +36,21 @@ export class Gateway extends GatewayStorage {
...
+	/**
+	 * The provider's name that manages this gateway.
+	 */
+	private readonly _provider: string;
+
+	public constructor(client: Client, name: string, options: GatewayOptions = {}) {
+		this.client = client;
+		this.name = name;
+		this.schema = options.schema || new Schema();
+		this._provider = options.provider || client.options.providers.default || '';
```
Comment (vladfrangu) [inline src/lib/settings/gateway/Gateway.ts:53]: Might as well use a private (`#`) field?

File: src/lib/settings/schema/Schema.ts [hunk 5]
```diff
@@ -1,24 +1,15 @@
-import { isFunction } from '@klasa/utils';
-import { SettingsFolder } from '../SettingsFolder';
+import { SchemaEntry, SchemaEntryOptions, SchemaEntryJson } from './SchemaEntry';
+import { Settings } from '../Settings';
+import { Cache } from '@klasa/cache';
```
Comment (vladfrangu) [inline src/lib/settings/schema/Schema.ts:3]:
```suggestion
import { Cache } from '@klasa/cache';
import { SchemaEntry, SchemaEntryOptions, SchemaEntryJson } from './SchemaEntry';
import { Settings } from '../Settings';
```

**Pull request merged**